pkg/allocator/cache.go

```diff
@@ -0,0 +1,229 @@
+	// master-key protection, if enabled, where the local allocator will recreate
+	// the key->ID association is recreated because the local node is still using
+	// it.
+	OnDelete(id idpool.ID, key AllocatorKey)
+}
+
+func (c *cache) sendEvent(typ kvstore.EventType, id idpool.ID, key AllocatorKey) {
+	if events := c.allocator.events; events != nil {
+		events <- AllocatorEvent{Typ: typ, ID: id, Key: key}
+	}
+}
+
+func (c *cache) OnListDone() {
+	c.mutex.Lock()
+	// nextCache is valid, point the live cache to it
+	c.cache = c.nextCache
+	c.keyCache = c.nextKeyCache
+	c.mutex.Unlock()
+
+	// report that the list operation has
+	// been completed and the allocator is
+	// ready to use
+	close(c.listDone)
+}
+
+func (c *cache) OnAdd(id idpool.ID, key AllocatorKey) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.nextCache[id] = key
+	if key != nil {
+		c.nextKeyCache[key.GetKey()] = id
+	}
+	c.allocator.idPool.Remove(id)
+
+	c.sendEvent(kvstore.EventTypeCreate, id, key)
+}
+
+func (c *cache) OnModify(id idpool.ID, key AllocatorKey) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if k, ok := c.nextCache[id]; ok {
+		delete(c.nextKeyCache, k.GetKey())
+	}
+
+	c.nextCache[id] = key
+	if key != nil {
+		c.nextKeyCache[key.GetKey()] = id
+	}
+
+	c.sendEvent(kvstore.EventTypeModify, id, key)
+}
+
+func (c *cache) OnDelete(id idpool.ID, key AllocatorKey) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	a := c.allocator
+	if a.enableMasterKeyProtection {
+		if value := a.localKeys.lookupID(id); value != nil {
+			a.backend.UpdateKey(id, value, true)
+			return
+		}
+	}
+
+	if k, ok := c.nextCache[id]; ok && k != nil {
+		delete(c.nextKeyCache, k.GetKey())
+	}
+
+	delete(c.nextCache, id)
+	a.idPool.Insert(id)
+
+	c.sendEvent(kvstore.EventTypeDelete, id, key)
+}
+
+// start requests a LIST operation from the kvstore and starts watching the
+// prefix in a go subroutine.
+func (c *cache) start(a *Allocator) waitChan {
+	c.listDone = make(waitChan)
+
+	c.mutex.Lock()
+
+	// start with a fresh nextCache
+	c.nextCache = idMap{}
+	c.nextKeyCache = keyMap{}
+	c.mutex.Unlock()
+
+	c.stopWatchWg.Add(1)
+
+	go func() {
+		c.allocator.backend.ListAndWatch(c, c.stopChan)
+		c.stopWatchWg.Done()
+	}()
+
+	return c.listDone
+}
+
+func (c *cache) stop() {
+	close(c.stopChan)
+	c.stopWatchWg.Wait()
+}
+
+func (c *cache) get(key string) idpool.ID {
+	c.mutex.RLock()
+	if id, ok := c.keyCache[key]; ok {
+		c.mutex.RUnlock()
+		return id
+	}
+	c.mutex.RUnlock()
+
+	return idpool.NoID
+}
+
+func (c *cache) getByID(id idpool.ID) AllocatorKey {
+	c.mutex.RLock()
+	if v, ok := c.cache[id]; ok {
+		c.mutex.RUnlock()
+		return v
+	}
+	c.mutex.RUnlock()
+
+	return nil
+}
+
+func (c *cache) foreach(cb RangeFunc) {
+	c.mutex.RLock()
+	for k, v := range c.cache {
+		cb(k, v)
+	}
+	c.mutex.RUnlock()
+}
+
+func (c *cache) insert(key AllocatorKey, val idpool.ID) {
+	c.mutex.Lock()
+	c.nextCache[val] = key
+	c.nextKeyCache[key.GetKey()] = val
+	c.mutex.Unlock()
+}
```

Comment:
Are there any notable code changes in this file vs. just moving to the new package? Would be helpful to have them pointed out if so.

Reply:
Unfortunately, yes. I'll mark them above, but I'll also try to redo the PR with a commit to copy code between packages and another that creates the abstraction.